hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/operations/CustomSelfDestructOperation.java

```diff
@@ -92,65 +97,135 @@ public OperationResult execute(@NonNull final MessageFrame frame, @NonNull final
                     tbdAddress, beneficiaryAddress, inheritance.toLong(), isDelegateCall(frame));
-            if (maybeReasonToHalt.isPresent()) {
-                return new OperationResult(cost, maybeReasonToHalt.get());
-            }
+            if (maybeReasonToHalt.isPresent()) return resultFor(costWithoutBeneficiary, maybeReasonToHalt.get());
 
-            // Tell the EVM to delete this contract if pre-Cancun, or, if post-Cancun, only in the
-            // same transaction it was created in
-            final boolean tellEVMToDoContractDestruct =
-                    switch (eip6780Semantics) {
-                        case NO -> true;
-                        case YES -> frame.wasCreatedInTransaction(tbdAddress);
-                    };
+            // From this point success is assured ...
+
+            // Frame tracks contracts to be deleted (for handling later)
+            if (contractIsToBeDeleted) frame.addSelfDestruct(tbdAddress);
 
-            if (tellEVMToDoContractDestruct) {
-                frame.addSelfDestruct(tbdAddress);
+            if (!contractIsItsOwnBeneficiary || contractIsToBeDeleted) {
+                proxyWorldUpdater.trackSelfDestructBeneficiary(tbdAddress, beneficiaryAddress, frame);
+                // Frame tracks balance changes
+                frame.addRefund(beneficiaryAddress, inheritance);
             }
 
-            frame.addRefund(beneficiaryAddress, inheritance);
             frame.setState(State.CODE_SUCCESS);
-            return new OperationResult(cost, null);
+            return resultFor(costWithBeneficiary, null);
+
         } catch (UnderflowException ignore) {
             return UNDERFLOW_RESPONSE;
         }
     }
 
-    private OperationResult haltFor(
-            @Nullable final Account beneficiary, final long inheritance, @NonNull final ExceptionalHaltReason reason) {
-        final long cost = gasCalculator().selfDestructOperationGasCost(beneficiary, Wei.of(inheritance));
+    protected @NonNull Optional<ExceptionalHaltReason> validateHederaRestrictionsOnBeneficiary(
+            @NonNull final Address deleted, @NonNull final Address beneficiary, @NonNull final MessageFrame frame) {
+        requireNonNull(deleted);
+        requireNonNull(beneficiary);
+        requireNonNull(frame);
+
+        final var proxyWorldUpdater = (ProxyWorldUpdater) frame.getWorldUpdater();
+        final var beneficiaryAccount = proxyWorldUpdater.getAccount(beneficiary);
+
+        // Beneficiary must not be a system account, and ...
+        if (addressChecks.isSystemAccount(beneficiary)) return Optional.of(INVALID_SOLIDITY_ADDRESS);
+
+        // ... must be present in the frame, and ...
+        if (!addressChecks.isPresent(beneficiary, frame)) return Optional.of(INVALID_SOLIDITY_ADDRESS);
+
+        // must exist, and ...
+        if (beneficiaryAccount == null) return Optional.of(INVALID_SOLIDITY_ADDRESS);
+
+        // ... must not be a token or schedule.
```

Comment:
do we need to check to see if the beneficiary has been deleted?

Reply:
hmm.  probably!

Reply:
Turns out, if the account corresponds to a deleted Hedera account then `proxyWorldUpdater.getAccount` up at L185 would have returned `null` and thus we would have failed the validation already.